src/http/web_socket.cr

```diff
@@ -27,38 +35,66 @@ class HTTP::WebSocket
     @remaining = 0
   end
 
-  def send(data)
-    write_header(data.size)
-    @io.print data
-    @io.flush
-  end
+  class StreamIO
+    include IO
 
-  def send_masked(data)
-    write_header(data.size, true)
+    def initialize(@websocket, @opcode, @masked, frame_size)
+      @buffer = Slice(UInt8).new(frame_size)
+      @pos = 0
+    end
 
-    mask_array = StaticArray(UInt8, 4).new { rand(256).to_u8 }
-    @io.write mask_array.to_slice
+    def write(slice : Slice(UInt8))
+      count = Math.min(@buffer.size - @pos, slice.size)
+      (@buffer + @pos).copy_from(slice.pointer(count), count)
+      @pos += count
 
-    data.size.times do |index|
-      mask = mask_array[index % 4]
-      @io.write_byte (mask ^ data.byte_at(index).to_u8).to_u8
+      if @pos == @buffer.size
+        flush(final: false)
+      end
+
+      if count < slice.size
+        count += write(slice + count)
+      end
+
+      count
+    end
+
+    def read(slice : Slice(UInt8))
+      raise "this IO is write-only"
+    end
+
+    def flush(final = true)
+      @websocket.send(
+        @buffer + (@pos % @buffer.size),
+        @opcode,
+        @masked,
+        flags: final ? Flags::FINAL : Flags::None,
+        flush: final
+      )
+      @opcode = Opcode::CONTINUATION
+      @pos = 0
     end
-    @io.flush
   end
 
-  private def write_header(size, masked = false)
-    @io.write_byte(0x81_u8)
+  {% for data_type, opcode in { "String" => "Opcode::TEXT", "Slice(UInt8)" => "Opcode::BINARY" } %}
+    def send(data : {{data_type.id}}, masked = false)
+      slice = {{data_type == "Slice(UInt8)" ? "data".id : "data.to_slice".id}}
+      send(slice, {{opcode.id}}, masked)
+    end
+  {% end %}
 
-    mask = masked ? MASK_BIT : 0
-    if size <= 125
-      @io.write_byte(size.to_u8 | mask)
-    elsif size <= UInt16::MAX
-      @io.write_byte(126_u8 | mask)
-      1.downto(0) { |i| @io.write_byte((size >> i * 8).to_u8) }
-    else
-      @io.write_byte(127_u8 | mask)
-      3.downto(0) { |i| @io.write_byte((size >> i * 8).to_u8) }
+  {% for type, opcode in { "text" => "Opcode::TEXT", "binary" => "Opcode::BINARY" } %}
+    def {{type.id}}_stream(masked = false, frame_size = 1024)
+      stream_io = StreamIO.new(self, {{opcode.id}}, masked, frame_size)
+      yield(stream_io)
+      stream_io.flush
     end
+  {% end %}
```

Comment:
Instead of using macros, just define a `stream` method and add a named argument `binary = true`. Using that `binary` argument use the appropriate opcode.